src/main/java/se/liu/ida/hefquin/engine/data/impl/VocabularyMappingImpl.java

```diff
@@ -40,23 +54,293 @@ public VocabularyMappingImpl(final Set<Triple> triples) {
+		final Node p = NodeFactory.createVariable("p");
+		final Node o = NodeFactory.createVariable("o");
+		final TriplePattern tpQuery = new TriplePatternImpl(t.getSubject(), p, o);
+		for (final Triple m : getMappings(tpQuery)) {
+			if (m.asJenaTriple().getPredicate().getURI().equals(OWL.sameAs.getURI())) {
+				final TriplePattern translation = new TriplePatternImpl(m.asJenaTriple().getObject(), t.getPredicate(), t.getObject());
+				results.add(translation);
+			} else {
+				throw new IllegalArgumentException(m.asJenaTriple().getPredicate().getURI());
+			}
+		}
+		return results;
 	}
 	
 	protected SPARQLGraphPattern translateObject(final TriplePattern tp){
-		return null;
+		org.apache.jena.graph.Triple t = tp.asJenaTriple();
+		if(!t.getPredicate().isURI()) {
+			//System.out.print("Not URI : " + t.getPredicate().toString());
+			return tp;
+		} else {
+			if(!t.getPredicate().getURI().equals(RDF.type.getURI())) {
+				//System.out.print("Not RDF type : " + t.getPredicate().getURI());
+				return tp;
+			}
+		}
+		final Node p = NodeFactory.createVariable("p");
+		final Node o = NodeFactory.createVariable("o");
+		final TriplePattern tpQuery = new TriplePatternImpl(t.getObject(), p, o);		
+		List<SPARQLGraphPattern> resultsList = new ArrayList<SPARQLGraphPattern>();
+		for (final Triple m : getMappings(tpQuery)) {
+			String predicate = m.asJenaTriple().getPredicate().getURI();
+			
+			if(predicate.equals(OWL.sameAs.getURI())) {
+				final TriplePattern translation = new TriplePatternImpl(t.getSubject(), t.getPredicate(), m.asJenaTriple().getObject());
+				resultsList.add(translation);
+			} else if (predicate.equals(OWL.equivalentClass.getURI())) {
+				
+				/*
+				 * Blank node used to help create intersections and unions
+				 * (a equals b)
+				 * (b unionof c)
+				 * (b unionof d)
+				 */
+				if (m.asJenaTriple().getObject().isBlank()) {
+					//Union or intersection
+					final Node newP = NodeFactory.createVariable("p");
+					final Node newO = NodeFactory.createVariable("o");
+					final TriplePattern blankQuery = new TriplePatternImpl(m.asJenaTriple().getObject(), newP, newO);
+					String newPredicate = "";
+					List<Node> objects = new ArrayList<Node>();
+					for(final Triple i : getMappings(blankQuery)) {
+						String iPredicate = i.asJenaTriple().getPredicate().getURI();
+						if(newPredicate.equals("")) {
+							newPredicate = iPredicate;
+						} else {
+							if (!iPredicate.equals(newPredicate)) {
+								throw new IllegalArgumentException(iPredicate);
+							}
+						}
+						objects.add(i.asJenaTriple().getObject());
+					}
+					
+					if (newPredicate.equals(OWL.unionOf.getURI())) {
+						//SPARQLUnionPatternImpl union = new SPARQLUnionPatternImpl();
+						List<SPARQLGraphPattern> unionList = new ArrayList<SPARQLGraphPattern>();
+						for(Node j : objects){
+							final TriplePattern translation = new TriplePatternImpl(t.getSubject(), t.getPredicate(), j);
+							//union.addSubPattern(translation)
```

Comment:
Why exactly did that didn't work?? Of course, you need a semicolon here but, other than that, I don't see a reason why you cannot use this one directly.

Reply:
I get an Unsupported Operation Exception when using this. It is the same error that I get when trying to add to lists/set created with Collections.EmptySet or EmptyList.

Reply:
Ah! I see!! Try again now with the fix in commit https://github.com/LiUSemWeb/HeFQUIN/pull/162/commits/bec0d4d2e12d790648c18f928122edd7359e0839